tsl/src/chunk.c

```diff
@@ -1350,43 +1455,660 @@ relation_split_info_free(RelationSplitInfo *rsi, int ti_options)
+		 * to be split across the partitions by decompressing and
+		 * recompressing into sub-segments.
+		 */
+		HeapTuple tuple;
+		CompressionSettings *csettings =
+			ts_compression_settings_get_by_compress_relid(RelationGetRelid(scontext->rel));
+
+		tuple = ExecFetchSlotHeapTuple(slot, false, NULL);
+
+		RowDecompressor decompressor =
+			build_decompressor(slot->tts_tupleDescriptor, csp->noncompressed_tupdesc);
+
+		heap_deform_tuple(tuple,
+						  decompressor.in_desc,
+						  decompressor.compressed_datums,
+						  decompressor.compressed_is_nulls);
+
+		int nrows = decompress_batch(&decompressor);
+
+		/*
+		 * Initialize a compressor for each new partition.
+		 */
+		for (int i = 0; i < scontext->split_factor; i++)
+		{
+			RelationWriteState *rws = &scontext->rws[i];
+			row_compressor_init(&rws->compressor,
+								csettings,
+								csp->noncompressed_tupdesc,
+								RelationGetDescr(scontext->rws[i].targetrel));
+		}
+
+		/*
+		 * Route each decompressed tuple to its corresponding partition's
+		 * compressor.
+		 */
+		for (int i = 0; i < nrows; i++)
+		{
+			int routing_index = route_tuple(decompressor.decompressed_slots[i], scontext->sp);
+			Assert(routing_index == 0 || routing_index == 1);
+			RelationWriteState *rws = &scontext->rws[routing_index];
+			/*
+			 * Since we're splitting a segment, the new segments will be
+			 * ordered like the original segment. Also, there is no risk of
+			 * the segments getting too big since we are only making segments
+			 * smaller.
+			 */
+			row_compressor_append_ordered_slot(&rws->compressor,
+											   decompressor.decompressed_slots[i]);
+		}
+
+		row_decompressor_close(&decompressor);
+		scontext->rws_index = 0;
+
+		/*
+		 * Call this function again to return the sub-segments.
+		 */
+		return route_next_compressed_tuple(slot, scontext, routing_index);
+	}
+
+	/* Update tuple count stats for compressed data */
+	bool isnull;
+	Datum count = slot_getattr(slot, csp->attnum_count, &isnull);
+	scontext->rws[*routing_index].stats.tuples_in_segments += DatumGetInt32(count);
+
+	/*
+	 * The compressed tuple (segment) can be routed without splitting it.
+	 */
+	Assert(*routing_index >= 0 && *routing_index < scontext->split_factor);
+	scontext->rws_index = scontext->split_factor;
+
+	return ExecFetchSlotHeapTuple(slot, false, NULL);
+}
+
+static double
+copy_tuples_for_split(SplitContext *scontext)
+{
+	Relation srcrel = scontext->rel;
+	TupleTableSlot *srcslot;
+	MemoryContext oldcxt;
+	EState *estate;
+	ExprContext *econtext;
+	TableScanDesc scan;
+	SplitPoint *sp = scontext->sp;
+
+	estate = CreateExecutorState();
+
+	/* Create the tuple slot */
+	srcslot = table_slot_create(srcrel, NULL);
+
+	/*
+	 * Scan through the rows using SnapshotAny to see everything so that we
+	 * can transfer tuples that are deleted or updated but still visible to
+	 * concurrent transactions.
+	 */
+	scan = table_beginscan(srcrel, SnapshotAny, 0, NULL);
```

Comment:
Can a concurrent transaction reading from the chunk survive its split and continue reading? Seems unlikely if we delete the table, but I'm not sure. Normally DROP or TRUNCATE don't support reading the old versions and use locking instead, right?

Reply:
The chunk that gets split is not dropped. It's constraints are shrunk to make place for a new partition (chunk) and then we write all the data into a new heap for the existing chunk and the heap for the new chunk. In a final step the existing chunk gets a heap swap while the other (new) chunk is attached to the hypertable.